OpenRA.Mods.Common/Activities/UnloadCargo.cs

```diff
@@ -121,9 +121,12 @@ public override bool Tick(Actor self)
 					var move = actor.Trait<IMove>();
 					var pos = actor.Trait<IPositionable>();
 
-					actor.CancelActivity();
+					pos.SetPosition(self, exitSubCell.Value.First, exitSubCell.Value.Second);
 					pos.SetVisualPosition(actor, spawn);
-					actor.QueueActivity(move.MoveIntoWorld(actor, exitSubCell.Value.First, exitSubCell.Value.Second));
+					if (actor.CurrentActivity == null)
+						actor.QueueActivity(move.MoveIntoWorld(actor, exitSubCell.Value.First, exitSubCell.Value.Second));
```

Comment:
The problem at hand is really an implementation issue/detail of `EnterTransport`, so it would be a lot cleaner if we could fix the issue in `EnterTransport`/`Enter` instead of adding workarounds (especially uncommented ones!) in other parts of the code.

Experience shows us that external workarounds tend to be forgotten when the code they are working around is inevitably rewritten, causing confusion and headaches for people in the future who are trying to look back and work out what odd code snippets were originally trying to solve.

Reply:
One simple(?) option could be to change `OnEnterComplete` to return a bool: returning false keeps the current behaviour where it transitions to `EnterState.Exiting`,  returning true jumps directly to `EnterState.Complete`.

Reply:
This is not adding an external workaround but removing one.

`Enter` and all its derived classes always take care of exiting as well. This makes sense because everything that enters must at some point exit (or despawn) before it does anything else. Whether it has finished or cancelled, the `MoveIntoWorld` must always happen and cannot be interrupted. `EnterTransport` isn't (and should be) and exception. Anything that enters a transport must exit before it can do anything else. The only difference is that the place where it exits is not the same as where it entered.

Implementing complex exceptions in `Enter` is a bad idea: It makes the behaviour contradictory with every other use case of `Enter` and makes the activity even more convoluted than it already is. I predict that this approach is much more susceptible to unforeseen regressions.

We would also be shooting ourselves in the foot in the long run: The way the activity system is designed, we cannot inject an activity in the middle of a queue (nor should we feel the need to) and we must avoid messing with running activity queues whenever possible. This means it would be impossible to implement for example an autotransport (analogous with `AutoCarryall`) if we would go this route. On the basis of this PR we would only need to make the last `actor.CancelActivity` call optional to make that possible.

Furthermore, any cargo actor that we can unload must have either entered at some point or spawned there. So we only need to manually queue `MoveIntoWorld` in the latter case (which only happens with scripted transports or a hypothetical future `ProductionTransport` trait). `UnloadCargo` should ideally not touch the cargo's activity queue at all but simply drop the cargo into the world with a specified position and let the cargo take care of the exit animation itself in all cases. We could do that by letting `Mobile` queue a `MoveIntoWorld` activity upon creation (which would also clean op the `Production` traits somewhat). Modelling `UnloadCargo` this way would also allow us to merge `UnloadCargo` and `DeliverUnit` quite easily, because `DeliverUnit` already operates on that model.

So, TL;DR: I think my solution is more internally consistent and sustainable in the long run.

Reply:
This will work better if we can rename `EnterTransport` to make it clear that it also manages exit, and ideally also use the same activity for units spawned in the transport instead of having a separate code path for `MoveIntoWorld` here.